sklearn/ensemble/_hist_gradient_boosting/_predictor_bitset.pyx

```diff
@@ -0,0 +1,104 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: language_level=3
+# cython: nonecheck=False
+# distutils: language=c++
+
+from ._bitset cimport in_bitset
+from .common cimport BITSET_INNER_DTYPE_C
+from libc.limits cimport CHAR_BIT
+from libcpp.vector cimport vector
+
+
+cdef inline unsigned char in_vec_bitset(vector[BITSET_INNER_DTYPE_C]& bitset,
+                                        int value) nogil:
+    cdef:
+        unsigned int i1 = value // 32
+        unsigned int i2 = value % 32
+
+    if bitset.size() < i1 + 1:
+        return 0
+    return (bitset[i1] >> i2) & 1
+
+
+cdef inline void insert_vec_bitset(vector[BITSET_INNER_DTYPE_C]& bitset,  # OUT
+                                   int value) nogil:
+    cdef:
+        unsigned int i1 = value // 32
+        unsigned int i2 = value % 32
+
+    if bitset.size() < i1 + 1:
+        bitset.resize(i1 + 1, 0)
+    bitset[i1] |= (1 << i2)
+
+
+cdef class PredictorBitSet:
+    def __init__(self, list bin_thresholds,
+                 const unsigned char [:] is_categorical):
+        """Creates bitset for all known categories"""
+        if is_categorical is None or bin_thresholds is None:
+            return
+
+        cdef:
+            int i
+            X_DTYPE_C category
+            unsigned int i1
+            unsigned int i2
+
+        for i in range(is_categorical.shape[0]):
+            if is_categorical[i] == 0:
+                continue
+            for category in bin_thresholds[i]:
+                insert_vec_bitset(self.feature_idx_raw_cats[i],
+                                  <int>(category))
```

Comment:
I am confused by this. As far as I understand Category is a raw category, i.e. it could be 
any arbitrary value like 123456? I'm not sure this makes sense to use a bitset for these, especially considering that we only seem to use `feature_idx_raw_cats` in `is_known_category`.

Shouldn't we be able to just rely on a set instead of a map?

Reply:
I tried a std::set first and it was slow during predict time.

Reply:
`unordered_set` might be what you need here